### DISCUSSION TITLE
(왕초보)로그인한 캐릭터로 캐릭터 생성되도록 다음의 가이드를 참고하여 제작해 주세요

### DISCUSSION THREAD

**Discussion by kiuworld** [naverz/zepeto-studio-kor#1236] [5. Q&A]
안녕하세요 월드맵 심사가 거절되었는데 궁금한것이 있어 글을 남깁니다 ㅠㅠ

[로그인한 캐릭터로 캐릭터 생성되도록 다음의 가이드를 참고하여 제작해 주세요]
라고 답변을 받았는데 제페토 스튜디오의 가이드 예제를 봐도 이해가 어려워서

현재 Character.Controller 스크립트는 예제 가이드를 보고 이렇게 수정하였는데 

import { ZepetoScriptBehaviour } from 'ZEPETO.Script'
import { SpawnInfo, ZepetoPlayers, LocalPlayer, ZepetoCharacter } from 
'ZEPETO.Character.Controller'
import { WorldService } from 'ZEPETO.World';
 
export default class CharacterController extends ZepetoScriptBehaviour {
   Start() {        
    // (Editor를 통해 로그인한)UserId를 가져옵니다.  
    ZepetoPlayers.instance.CreatePlayerWithUserId(WorldService.userId,new SpawnInfo(), true);             
    ZepetoPlayers.instance.OnAddedLocalPlayer.AddListener(() => {            
        let _player : LocalPlayer = ZepetoPlayers.instance.LocalPlayer;        
    });    
}
}

이것이 싱글플레이 월드에서 캐릭터 생성하는걸로 알고있는데  저는 멀티플레이 월드를 만들고싶어
멀티플레이어 서버나 로직은 가이드 영상을 보고 따라해서 구현해놓았는데 

예제 2-2 ZepetoPlayer 에서
멀티 플레이 월드에서 Player 생성하기 라는 항목에 아래 스크립트를

OnJoinPlayer(sessionId: string, player: Player) {
 
   // (Editor를 통해 로그인한)UserId를 가져옵니다.
   const userId = WorldService.userId;
   const isLocal = this.room.SessionId === player.sessionId;
 
   // SessionId를 UniqueId로 하여, ZEPETO 캐릭터를 생성합니다.
   ZepetoPlayers.instance.CreatePlayerWithUserId(sessionId, userId, new SpawnInfo(), isLocal);
}
 
OnLeavePlayer(sessionId: string, player: Player) {
       
   // SessionId를 UniqueId로 하여, ZEPETO 캐릭터를 제거합니다.
   ZepetoPlayers.instance.RemovePlayer(sessionId);
}

 Character.Controller 스크립트에 넣는건지 아니면 어디다 작성해야하는지 몰라서 문의 남깁니다 ㅠㅠ

**Comment by zlmrzlo** [comment]
https://github.com/naverz/zepeto-studio-kor/discussions/835
이것을 보시면 도움이 되실 것 같네요!

**Reply by kiuworld** [reply]
위 client starter 스크립트는 가이드 예제를 보고 미리 만들었는데 
https://studio.zepeto.me/guides/zepetoplayer
에 멀티 플레이 월드에서 Player 생성하기에 스크립트를 어디다 넣어줘야하는지를 모르겠습니다 ㅠㅠ

**Reply by zlmrzlo** [reply]
index.ts 이 파일에 넣으시면 될 것 같아요.

**Reply by zlmrzlo** [reply]
835를 보고 먼저 잘 돌아가는 프로젝트로 공부를 하시면 더 도움일 될 것 같아요. 씹고 뜯고 맛보고 즐기다 보면 잘하게 될 거에요!

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요, 답변을 드리자면 ClientStarter 내에 해당 코드를 구현하시면 됩니다.

싱글플레이든 멀티플레이든 씬에 제페토 플레이어 인스턴스를 로드하기 위해서는 
클라이언트 스크립트에 제페토 플레이어 생성 스크립트가 추가되어야 합니다.

싱글플레이 월드의 경우 멀티플레이 요소가 없기 때문에 모든 스크립트가 클라이언트 스크립트가 되는 것이고

멀티플레이의 경우는 World.multiplay > index.ts를 서버스크립트,
서버와 통신하는 클라이언트 스크립트를 샘플 예제에서는 ClientStarter라는 이름으로 제공하고 있습니다. 

유저가 월드에 들어오면 제일 먼저 서버에서 접속한 유저의 session id 정보를 받아서 
https://github.com/naverz/zepeto-multiplay-example/blob/main/Assets/ZepetoScripts/ClientStarterV2.ts
위 스크립트의 line 87 onJoinPlayer함수에 인자값으로 넘겨줍니다.

ZepetoPlayers.instance.CreatePlayerWithUserId(sessionId, player.zepetoUserId, spawnInfo, isLocal);
를 통해 제페토 플레이어 인스턴스를 생성하게 됩니다.

한국어로된 영상 강의도 있으니 참고 부탁 드립니다.
https://studio.zepeto.me/guides/multiplay-tutorial
https://youtu.be/tfkY_raboV0

감사합니다.

**Reply by kiuworld** [reply]
아하 그럼 character controller 스크립트는

import { ZepetoScriptBehaviour } from 'ZEPETO.Script'
import { SpawnInfo, ZepetoPlayers, LocalPlayer, ZepetoCharacter } from
'ZEPETO.Character.Controller'
import { WorldService } from 'ZEPETO.World';

export default class CharacterController extends ZepetoScriptBehaviour {
Start() {
// (Editor를 통해 로그인한)UserId를 가져옵니다.
ZepetoPlayers.instance.CreatePlayerWithUserId(WorldService.userId,new SpawnInfo(), true);
ZepetoPlayers.instance.OnAddedLocalPlayer.AddListener(() => {
let _player : LocalPlayer = ZepetoPlayers.instance.LocalPlayer;
});
}
}

이상태로 놔두어도 괜찮은건가요?

**Reply by kiuworld** [reply]
![image](https://user-images.githubusercontent.com/111574894/187324908-9be95ab8-6e29-492f-a070-ba77ce7ca9bb.png)

플레이 버튼을 누를시 이렇게 캐릭터가 복제되는데 정상인건가요?

**Reply by zlmrzlo** [reply]
정상이에요. 월드에서 캐릭터를 불러올 때 제페토 앱에서 캐릭터가 입던 옷으로 불러옵니다.

**Reply by Zepeto-TechEvangelist** [reply]
멀티플레이의 경우 character controller 스크립트는 사용하지 않으셔도 됩니다. 
ZepetoPlayers.instance.CreatePlayerWithUserId(WorldService.userId,new SpawnInfo(), true); 이 코드가 제페토 플레이어 인스턴스를 생성하기 때문에
이 코드가 스크립트 두 개에 있으니 캐릭터 두 개가 생기는 것입니다.

**Reply by kiuworld** [reply]
확인해서 해결했습니다 감사합니다!!
